### ISSUE TITLE
Implement `share` operation

### ISSUE THREAD

**Issue by mccorby** [OpenMined/AndroidWorker#24]
Implement the `share` operation in JVM/Android.
This operation will require the server side (PySyft) to work with tensors represented with `internal_type=torch.int16` in order to avoid the issues with the precision in the JVM.

**Comment by mccorby** [comment]
This is becoming a blocker. Since we are using DL4J as the framework and DL4J transforms all tensors into float, when working with shares (that are transformed into integers) there is a loss of precision at the moment of recreating the original tensor.

**Comment by karlhigley** [comment]
Could this potentially be addressed by integrating PyTorch mobile as a replacement for DL4J?

**Comment by mccorby** [comment]
It could when PyTorch mobile allowed training. I think right now it only does inference by loading a static model, sort of the same limitation with TensorFlow Lite.

**Comment by karlhigley** [comment]
I spent a bunch of time yesterday trying to find a viable cross-platform option for training on device, and the best approach I could find was interfacing with C++ libtorch through JNI on Android and ObjC on iOS. @iamtrask tells me that we don't need the underlying framework to support training on device though, because PySyft knows how to translate autograd into simple tensor ops.

**Comment by mccorby** [comment]
`PySyft knows how to translate autograd into simple tensor ops.`: If that's the case (and I don't doubt it is) then we might be able of using other libraries (I've been testing KMath). A new world is open in front of us :)

**Comment by karlhigley** [comment]
Seems like there may be some value in using the same library on Android and iOS. If we can build for both mobile platforms using libtorch (or another cross-platform alternative), might be easier to keep feature parity between the worker libraries.

**Comment by karlhigley** [comment]
> PySyft knows how to translate autograd into simple tensor ops

Let me revise that: what @iamtrask actually said was that *we* know how to translate autograd into simple tensor ops. I don't think `AutogradTensor` supports it yet; if I'm not mistaken, [this issue](https://github.com/OpenMined/PySyft/issues/2797) is related to fixing that.
